controllers/Handle.php

```diff
@@ -150,21 +168,53 @@ public function callback(string $provider): RedirectResponse
     public function redirect(string $provider): RedirectResponse
     {
         if (!in_array($provider, $this->enabledProviders)) {
-            abort(404);
+            return $this->redirectToSigninPage('This provider is not enabled');
         }
 
         if ($this->authManager->getUser()) {
             // @TODO:
             // - Handle case of user explicitly attaching a SSO provider to their account
             // - Localization
-            Flash::error("You are already logged in. Please log out first.");
+            Flash::error('You are already logged in. Please log out first.');
             return Redirect::back();
         }
 
         $config = Config::get('services.' . $provider, []);
 
-        return Socialite::driver($provider)
-            ->scopes($config['scopes'] ?? [])
-            ->redirect();
+        try {
+            $response = Socialite::with($provider)
+                ->scopes($config['scopes'] ?? [])
+                ->redirect();
+        } catch (Exception $e) {
+            return $this->redirectToSigninPage($e->getMessage());
+        }
+        return $response;
+    }
+
+    /*
+     * Returns canonical form for google emails.
+     * Remove +specifier after any email username
+     */
+    public function normalizeEmail($email)
+    {
+        [$user, $domain] = explode('@', $email);
+
+        if (in_array($domain, ['gmail.com', 'googlemail.com'])) {
+            // Google emails can have "." anywhere in the username but the actual account has none.
+            $user = str_replace('.', '', $user);
+        }
+        # user+specifier@domain
+        # remove "+specifier" for all email accounts.
+        $user = preg_replace('#\+.+#', '', $user);
```

Comment:
This seems dangerous, why is this required?

Reply:
@LukeTowers Why "dangerous" ?

In an email address, you can have anything after the + e.g. luke+test1@wintercms.com, luke+test2@wintercms.com, etc and they all deliver emails to luke@wintercms.com in the end. So it's better to remove those "extension" in order to have a "normalized" email address to work with.

Reply:
@mjauvin "Plus Addressing" is a feature that was specific to gmail, only recently (2020 or 2022?) added to MS 365; and is not part of any standards. https://mozilla.fandom.com/wiki/User:Me_at_work/plushaters

It's not safe to assume that just because there's a plus in the address "Plus Addressing" is being used and also you can have as many plus characters as you want in the local part.

The email is what the email is; why are we trying to "normalize" it beyond lowercasing it? (note that technically according to standards the local part could still be interpreted by mail servers as case sensitive if they wanted to, but realistically you're fine to treat the whole email as case insensitive: https://stackoverflow.com/a/9808332).

Reply:
Might not be a standard, but I've used it for 25 years with postfix mail server and it's pretty well reckognised all over.

I can drop this if you prefer, or we could have an option that would do a more strict normalization.

Reply:
I guess at this point I'm not understanding what the value of "normalizing" this is, we should probably be fairly strict on the email matching logic when it comes to authentication lest we open ourselves up to potential attacks.

Reply:
ok, I'm leaving the plus addressing alone.

Reply:
@mjauvin for what it's worth, I did a search up on whether plus aliases for emails have any ramifications for SSO purposes, and as far as SSO providers are concerned, there's no security issue with using them. Ultimately, it's up to the SSO provider whether they accept it or not (or handle it differently) - we shouldn't be stripping or otherwise modifying the email address when sending it to an SSO provider for authorisation.

Reply:
that was done after receiving the email address from the SSO provider, before trying to match it against our db

Reply:
Ah gotcha - I will just close me mouth then. :stuck_out_tongue: